pkg/webhook/mutatingwebhook/mutate.go

```diff
@@ -19,51 +20,60 @@ var (
 	deserializer  = codecs.UniversalDeserializer()
 )
 
-type mutateHandler func(admReview v1.AdmissionReview) *v1.AdmissionResponse
+type mutateHandler func(admReview admissionv1.AdmissionReview) *admissionv1.AdmissionResponse
 
+// handleMutate is a common function that decodes an admission review request before handing it off to the
+// mutator for processing and then writes the response
 func handleMutate(logger logr.Logger, w http.ResponseWriter, r *http.Request, mutator mutateHandler) {
-	var respBody []byte
-	body, err := io.ReadAll(r.Body)
+	admReviewBody, err := io.ReadAll(r.Body)
 	defer func() {
 		if err := r.Body.Close(); err != nil {
 			logger.Error(err, "unable to close the body")
 		}
 	}()
 	if err != nil {
-		logger.Error(err, "unable to read the body of the request")
-		w.WriteHeader(http.StatusInternalServerError)
-		respBody = []byte("unable to read the body of the request")
-	} else {
-		// mutate the request
-		respBody = mutate(logger, body, mutator)
-		w.WriteHeader(http.StatusOK)
+		msg := "unable to read the body of the request"
+		logger.Error(err, msg)
+		writeResponse(logger, http.StatusInternalServerError, w, []byte(msg))
+		return
 	}
-	if _, err := io.WriteString(w, string(respBody)); err != nil {
-		logger.Error(err, "unable to write response")
-	}
-}
 
-func mutate(logger logr.Logger, body []byte, mutator mutateHandler) []byte {
-	admReview := v1.AdmissionReview{}
-	if _, _, err := deserializer.Decode(body, nil, &admReview); err != nil {
-		logger.Error(err, "unable to deserialize the admission review object", "body", string(body))
-		admReview.Response = responseWithError(err)
+	// deserialize the request
+	admReview := admissionv1.AdmissionReview{}
+	if _, _, err := deserializer.Decode(admReviewBody, nil, &admReview); err != nil {
+		logger.Error(err, "unable to deserialize the admission review object", "body", string(admReviewBody))
+		writeResponse(logger, http.StatusBadRequest, w, []byte("unable to read the body of the request"))
```

Comment:
I'm curious why you just write an http response here and return instead of using an admission response with not-allowed and an error?

Reply:
It's because the AdmissionResponse is added to `admReview.Response` then when we write the response the body is the full `admReview` object and until line 55 we haven't really validated `admReview`. I'm not sure if you can return the AdmissionResponse object by itself if that's what you're suggesting. I haven't tried that. The original code (before I added the vm mutator) returned the full `admReview` object too.

What I have:
```
// sets "Response"
admReview.Response = mutator(admReview)

// respBody is the full admReview
respBody, err := json.Marshal(admReview)
if err != nil {
	logger.Error(err, "unable to marshal the admission review with response", "admissionReview", admReview)
	writeResponse(logger, http.StatusInternalServerError, w, []byte("failed to marshal the adm review resposne"))
	return
}

// the full respBody is written as the response
writeResponse(logger, http.StatusOK, w, respBody)
```

Original code:
https://github.com/codeready-toolchain/member-operator/blob/fdca75cbca46b4d2edd41c4fef5eeb05a2704a86/pkg/webhook/mutatingwebhook/mutate.go#L82-L86

Looking at the original code the full `admReview` is returned but what's interesting is it set admReview.Response even if the decoding fails. I'm not sure that's the correct thing to do though.

Reply:
As discussed, I changed it to use `admReview.Response` in the error cases for readability and since that's how it was in the original code. See c1f6dff0eeeab396c4031454704a063a519dacd8